File: u1/Sample_Code/SharedDataTest.java

public class SharedDataTest {

    public static void main(String args[]) {
        if (args.length != 2) {
            System.out.println("Correct usage: java SharedDataTest numThreads reps");
            System.exit(1);
        }
        int numThreads = Integer.parseInt(args[0]);
        int reps = Integer.parseInt(args[1]);
        int correctSum = numThreads * reps;
        Buf buf = new Buf();
        Thread[] myThread;
        myThread = new Thread[numThreads];
        for (int i = 0; i < numThreads; i++) {
            myThread[i] = new Thread(new Adder(reps, buf));
        }
        for (int i = 0; i < numThreads; i++) {
            myThread[i].start();
        }
        try {
            for (int i = 0; i < numThreads; i++) {
                myThread[i].join();
            }
        } catch (InterruptedException ie) {
            System.out.println("Main thread cannot join other threads.");
        }
        System.out.println("Computed sum = " + buf.getSum());
        System.out.println("Correct sum = " + (numThreads * reps));
        System.out.println("Overlapped increments = "
                + 100.0 * (correctSum - buf.getSum()) / correctSum + "%");
    }
}

// class for common shared buffer
class Buf {
    private int sum;
    
    public Buf() {
        sum = 0;
    }
    
    public int getSum() {
        return sum;
    }

    // method 1: add synchronized to make it thread safe
    //synchronized public void increment()
    public void increment() 
    {
        int tmpSum = sum;
        System.out.println("Thread-" + Thread.currentThread().getId() + "\tread " + tmpSum);
        sum = tmpSum + 1;
    }   
}

class Adder implements Runnable {

    private int repetitions;
    Buf myBuf;

    public Adder(int count, Buf buf) {
        repetitions = count;
        myBuf = buf;
    }
    public void run() {
        for (int i = 0; i < repetitions; i++) {
            // method 2: use lock in myBuf
            // synchronized(myBuf) 
            {
                myBuf.increment();
            }
        }
    }
}

File: u3/Python/SharedDataTest.py
import sys
import threading

class Adder(threading.Thread):
    """Adder"""
    
    def __init__(self, buf, cnt):
        threading.Thread.__init__(self)
        self.buf = buf
        self.cnt = cnt
    
    def run(self):
        for count in range(0,self.cnt):
            self.buf.increment()
            
class Buf:
    """Buf"""
    
    def __init__(self):
        self.sum = 0
        
    def getSum(self):
        return self.sum 
        
    def increment(self):
        # non atomic operation for the modification of common value
        tmpSum = self.sum
        # report thread index and global value before modification
        print('Thread-%d\tread %d' % (threading.get_ident(),tmpSum))
        self.sum = tmpSum + 1

if __name__ == "__main__":
    # Get arguments from command line
    if len(sys.argv) != 3:
        print('Correct usage: python SharedDataTest.py numThreads reps')
        sys.exit(1)

    numThreads = int(sys.argv[1])
    reps = int(sys.argv[2])

    # create shared Buf instance
    buf = Buf()
    # create list of thread instance
    tlist = [Adder(buf,reps) for i in range (0,numThreads)]

    # start each thread
    for t in tlist:
        t.start()

    # join each thread
    for t in tlist:
        t.join()

    # generate report
    correctSum = numThreads * reps
    print ('Computed sum = %d' % (buf.getSum()))
    print ('Correct sum = %d' % (correctSum))
    print ('Overlapped increments = %f%%' % (100.0 * (correctSum - buf.getSum()) / correctSum))